### PR TITLE
L2 Tokenisation

### PR DESCRIPTION
**Pull request by RogerTaule**
WIP

### CODE REVIEW

**Review by imagobea — COMMENTED** [review]
_Doncs molt be_ Tokenisation team! (ie well done) Hats off to you :tada: 

I just left some notes for myself and a very small number of suggestions where I've tagged Mr @RogerTaule 

Look fw to merge this and see how the "nightfall-node" implementation responds :crossed_fingers: Congrats again!!! :rocket:

**Review by daveroga — COMMENTED** [review]
First of all congrats for the amazing work done in this PR to all the team involved in the different parts and specially to @RogerTaule for the big effort in moving all the circuits to Circom and the different optimizations.

I only have left some comments to understand better some parts.

File: test/e2e/tokens/erc1155.test.mjs [hunk 1]
```diff
@@ -93,7 +71,7 @@ describe('ERC1155 tests', () => {
 
     await nf3Users[0].deposit(erc20Address, tokenType, transferValue, tokenId, 0);
 
-    await emptyL2();
+    await emptyL2(nf3Users[0], web3Client, eventLogs);
```
Comment (signorecello) [inline test/e2e/tokens/erc1155.test.mjs:74]: Can we pass this as an object? `emptyL2({user: nf3Users[0], web3: web3Client, logs: eventLogs})` and destructure them on the other side? So it's easier to see what does the method expect without having to check the other file

File: README.md [hunk 2]
```diff
@@ -64,9 +64,9 @@ details on how to do the latter as there are some additional considerations.
   - Use `-g` to use a Ganache client inside the container
   - Use `-l` to use some localhost client running on your machine. We recommend using Ganache first
     to check everything works, because it's considerably faster.
-- Additionally, you can use the `-s` flag. If you do that, Nightfall_3 will run with stubbed ZKP
-  circuits, which generate proofs that always verify. That's useful for development work because
-  tests will run much faster but clearly you should run without stubs, as a final check.
+  - Use `-r` to use ropsten node hosted by the dev team. Note: with option -r set environment
```
Comment (signorecello) [inline README.md:67]: I think we don't have ropsten anymore?

**Review by signorecello — CHANGES_REQUESTED** [review]
I can't even grasp the amount of work this is. Mad respect by the three of you that have written such good stuff

File: nightfall-optimist/src/services/block-utils.mjs [hunk 3]
```diff
@@ -1,14 +1,24 @@
...
 // These functions are called by static methods in the Block class but are sometimes needed when the rest
 // of the block object isn't.  They can thus be called directly when instantiating the Block class
 // would be problematic because of its reliance on the Optimist database.
 
-const { SIGNATURES } = config;
+const { SIGNATURES, TIMBER_HEIGHT } = config;
 const { ZERO } = constants;
+const { generalise } = gen;
+
+export const packInfo = (blockNumberL2, leafCount, proposer) => {
```
[the remark this answers is not in the document]
Comment (RogerTaule) [reply nightfall-optimist/src/services/block-utils.mjs:15]: We have different packing, one for the transaction structure and one for the block one. However, it is true that there was code duplicated, I'll try to remove it

File: docker/docker-compose.adversary.yml [hunk 4]
```diff
@@ -31,7 +29,7 @@ services:
     environment:
       ENVIRONMENT: ${ENVIRONMENT:-localhost}
       WEBSOCKET_PORT: ${WEBSOCKET_PORT:-8080}
-      BLOCKCHAIN_WS_HOST: ${BLOCKCHAIN_WS_HOST:-host.docker.internal}
+      BLOCKCHAIN_WS_HOST: ${BLOCKCHAIN_WS_HOST:-blockchain}
```
[the remark this answers is not in the document]
Comment (druiz0992) [reply docker/docker-compose.adversary.yml:32]: We are using standalone.geth for some tests. I dont care if adversary uses ganache, but please dont remove the yml

**Review by daveroga — APPROVED** [review]

**Pull request merged**